File: app/src/main/java/com/example/hsrrelicmanager/RelicAdapter.java
package com.example.hsrrelicmanager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class RelicAdapter extends RecyclerView.Adapter<RelicAdapter.ViewHolder> {

    private List<Relic> relicData;

    public RelicAdapter(List<Relic> relicData) {
        this.relicData = relicData;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.inventory_relic_item, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Relic relic = relicData.get(position);

        holder.getTvSet().setText(relic.getSet());
        holder.getTvMainstat().setText(relic.getMainstat());
        holder.getImgSet().setImageResource(relic.getImage());

        int backgroundResource = getBackgroundResource(relic);
        holder.getImgSet().setBackgroundResource(backgroundResource);

        Iterator<Map.Entry<String, Double>> itSubstat = relic.getSubstats().entrySet().iterator();
        Iterator<TextView> itTextView = holder.getTvSubstats().iterator();
        Iterator<TextView> itTextViewVal= holder.getTvSubstatVals().iterator();

        while (itSubstat.hasNext() && itTextView.hasNext() && itTextViewVal.hasNext()) {
            Map.Entry<String, Double> substat = itSubstat.next();
            TextView textView = itTextView.next();
            TextView textViewVal = itTextViewVal.next();

            textView.setText(substat.getKey() + ": ");
            textViewVal.setText(substat.getValue().toString());
        }
    }

    @Override
    public int getItemCount() {
        return relicData.size();
    }

    private int getBackgroundResource(Relic relic) {
        if (relic.getRarity() == 1) {
            return R.drawable.bg_1_star;
        } else if (relic.getRarity() == 2) {
            return R.drawable.bg_2_star;
        } else if (relic.getRarity() == 3) {
            return R.drawable.bg_3_star;
        } else if (relic.getRarity() == 4) {
            return R.drawable.bg_4_star;
        } else {
            return R.drawable.bg_5_star;
        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvSet, tvMainstat, tvSubstat1, tvSubstat2, tvSubstat3, tvSubstat4,
                tvMainstatVal, tvSubstat1Val, tvSubstat2Val, tvSubstat3Val, tvSubstat4Val;
        private ImageView imgSet;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            tvSet = itemView.findViewById(R.id.tvSet);
            tvMainstat = itemView.findViewById(R.id.tvMainstat);
            tvSubstat1 = itemView.findViewById(R.id.tvSubstat1);
            tvSubstat2 = itemView.findViewById(R.id.tvSubstat2);
            tvSubstat3 = itemView.findViewById(R.id.tvSubstat3);
            tvSubstat4 = itemView.findViewById(R.id.tvSubstat4);

            tvMainstatVal = itemView.findViewById(R.id.tvMainstatVal);
            tvSubstat1Val = itemView.findViewById(R.id.tvSubstat1Val);
            tvSubstat2Val = itemView.findViewById(R.id.tvSubstat2Val);
            tvSubstat3Val = itemView.findViewById(R.id.tvSubstat3Val);
            tvSubstat4Val = itemView.findViewById(R.id.tvSubstat4Val);

            imgSet = itemView.findViewById(R.id.imgSet);
        }

        public TextView getTvSet() {
            return tvSet;
        }

        public TextView getTvMainstat() {
            return tvMainstat;
        }

        public TextView getTvMainstatVal() {
            return tvMainstatVal;
        }

        public List<TextView> getTvSubstats() {
            return Arrays.asList(tvSubstat1, tvSubstat2, tvSubstat3, tvSubstat4);
        }

        public List<TextView> getTvSubstatVals() {
            return Arrays.asList(tvSubstat1Val, tvSubstat2Val, tvSubstat3Val, tvSubstat4Val);
        }

        public ImageView getImgSet() {
            return imgSet;
        }
    }
}

File: app/src/main/java/com/example/hsrrelicmanager/RelicAdapter.kt
package com.example.hsrrelicmanager

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.example.hsrrelicmanager.databinding.InventoryRelicItemBinding
import com.example.hsrrelicmanager.model.relics.Relic

class RelicAdapter(
    private val relicData: MutableList<Relic>,
    private val inventoryBodyFragment: InventoryBodyFragment,
    private var selectedPos: Int = RecyclerView.NO_POSITION
) :
    RecyclerView.Adapter<RelicAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = InventoryRelicItemBinding.inflate(LayoutInflater.from(parent.context))
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val relic = relicData[position]
        holder.bind(relic, position)
    }

    override fun getItemCount(): Int {
        return relicData.size
    }

    inner class ViewHolder(val binding: InventoryRelicItemBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(relic: Relic, position: Int) {
            binding.apply {
                lblRelicName.text = relic.set.name
                lblRelicMainStatType.text = relic.mainstat
                lblRelicMainStatValue.text = relic.mainstatVal
                imgRelicMainStat.setImageResource(relic.mainstatResource)

                imgRelic.setImageResource(relic.set.icon)
                if (relic.prev != null && relic.level != relic.prev!!.level) {
                    lblRelicLevel.text = "+${relic.prev!!.level}  >  +${relic.level}"
                } else {
                    lblRelicLevel.text = "+${relic.level}"
                }

                imgRelic.setBackgroundResource(relic.rarityResource)

                val icons = listOf(
                    imgRelicSubStat1,
                    imgRelicSubStat2,
                    imgRelicSubStat3,
                    imgRelicSubStat4
                )
                val types = listOf(
                    lblRelicSubStatType1,
                    lblRelicSubStatType2,
                    lblRelicSubStatType3,
                    lblRelicSubStatType4
                )
                val values = listOf(
                    lblRelicSubStatValue1,
                    lblRelicSubStatValue2,
                    lblRelicSubStatValue3,
                    lblRelicSubStatValue4
                )
                val containers = listOf(
                    relicSubstatContainer1,
                    relicSubstatContainer2,
                    relicSubstatContainer3,
                    relicSubstatContainer4
                )

                for (i in containers.indices) {
                    if (i < relic.substats.size)
                        containers[i].visibility = View.VISIBLE
                    else
                        containers[i].visibility = View.INVISIBLE
                }

                val entries = relic.substats.entries
                for (i in entries.indices) {
                    val substat = entries.elementAt(i)
                    icons[i].setImageResource(relic.substatResource(substat.key))
                    types[i].text = substat.key
                    values[i].text = substat.value
                }

                if (relic.status.contains(Relic.Status.EQUIPPED)) {
                    containerRelicEquipped.visibility = View.VISIBLE
                } else {
                    containerRelicEquipped.visibility = View.INVISIBLE
                }

                val relics = relic.status.filter {
                    it != Relic.Status.EQUIPPED
                }.map {
                    relic.statusResource(it)
                }
                when (relics.size) {
                    0 -> {
                        containerSingleRelicStatus.visibility = View.INVISIBLE
                        containerDoubleRelicStatus.visibility = View.INVISIBLE
                    }
                    1 -> {
                        containerSingleRelicStatus.visibility = View.VISIBLE
                        containerDoubleRelicStatus.visibility = View.INVISIBLE
                        imgRelicStatus.setImageResource(relics[0])
                    }
                    2 -> {
                        containerSingleRelicStatus.visibility = View.INVISIBLE
                        containerDoubleRelicStatus.visibility = View.VISIBLE
                        imgRelicStatus1.setImageResource(relics[0])
                        imgRelicStatus2.setImageResource(relics[1])

                    }
                }
            }

            val relicCard = binding.root.findViewById<LinearLayout>(R.id.relic_card)

            if (position == selectedPos) {
                relicCard.setBackgroundResource(R.drawable.bg_inventory_relic_item_selected)
            } else {
                relicCard.setBackgroundResource(R.drawable.bg_inventory_relic_item)
            }

            binding.root.setOnClickListener {
                notifyItemChanged(selectedPos)

                selectedPos = if (selectedPos == position) {
                    RecyclerView.NO_POSITION
                } else {
                    position
                }

                notifyItemChanged(selectedPos)

                val relicBottomSheetFragment = RelicBottomSheetFragment(relic)
                relicBottomSheetFragment.show(inventoryBodyFragment.childFragmentManager, relicBottomSheetFragment.tag)
                inventoryBodyFragment.requireActivity().supportFragmentManager.setFragmentResultListener("relic", inventoryBodyFragment.viewLifecycleOwner) { _, bundle ->
                    val r = bundle.getParcelable<Relic>("relic")!!
                    relicData[position] = r
                    notifyItemChanged(position)
                }
            }
        }
    }
}